optionally ignore mouse click events associated with touch screen click event

derammo opened the pull request:
suppression of mouse after touch click is a configurable profile option for each monitor, affecting all the controls on that monitor
configuration parameter sets number of milliseconds to ignore mouse down and mouse up after a touch down event on the same control, with 0 being disabled
this works around problems with a touch screen that generates mouse events for each touch
event, resulting in triggering controls twice for every touch

this problem was observed with an ASUS VT168H using default Windows 10 drivers and configuration

BlueFinBima commented:
Hi @derammo ,  I really appreciate your time to write this feature and submit this pull request.  It is great to have new people exploring this code. 

The problem that you seem to be circumventing has been encountered by many, myself included, and I believe it is corrected by changing accessibility options in the settings of windows.  Unfortunately the settings go by many names in different releases of windows, and I can't recall exactly which one I personally needed to use (I think it was probably Double tap = Double Click (Win 8.1) in "Pen & Touch").  I apologise for being vague - the original post I saw on ED was also vague - but every person who reported this problem has found a solution in Windows.  

Did you try altering windows options before attempting to fix this with code?  If these did not work for you, I would be interested in knowing why you didn't feel a windows fix was adequate. 

I'm interested in your experiences.

I've temporarily pulled your code into a side branch to take a look.  

Cheers, Neil

derammo commented:
I have Windows 10 and the setting I found that has "Double tap = Double click" seems to be a fixed setting.  At least it did not appear to me that you can change it to do anything else.  It just has "how fast you want to click to count as a double click" in its settings but I didn't find how to turn it off.

Someone on Eagle forum said the same thing as you did, that I should be able to change this setting.  But I could not get it to work and also the sequence of events delivered varies greatly by tap.  Here is the data I posted about it on the forum:

> Update: I added some debug code to Helios and confirmed that when I press on my touch screen, it delivers the OnTouchDown event FOLLOWED by a spurious OnMouseDown event. Both of those fire the MouseDown (which becomes 'pushed' event I guess.)


Code:
```
Three mouse clicks:

OnMouseDown 25,17
OnMouseUp 25,17
OnMouseDown 21,19
OnMouseUp 21,19
OnMouseDown 19,25
OnMouseUp 19,25

Several touches, with various click speeds:

OnTouchDown 14,14
OnTouchUp 17,13

OnTouchDown 15,11
OnMouseDown 15,11 (seems like this is holding mouse down until two taps later, some accessibility thing?)
OnTouchUp 22,15   (maybe I slid here?)

OnTouchDown 18,8
OnTouchUp 18,8

OnTouchDown 28,16
OnTouchUp 30,13
OnMouseUp 30,-57 (finally released mouse) 

OnTouchDown 22,20
OnTouchUp 23,21

OnTouchDown 18,11
OnMouseDown 18,11
OnTouchUp 18,11
OnMouseUp 18,11 (this is a pretty normal tap)

OnTouchDown 17,11
OnMouseDown 17,11
OnTouchUp 17,11
OnMouseUp 17,11 (another normal one)
```

You see it is doing it some of the time, even though I am MOST DEFINITELY not double tapping.  I am just barely tapping the screen, and it is sometimes registering as double, and once above you can even see it seems to have thought I moved my finger?   I suspect the screen is deforming slightly or the sensor just sucks on this screen.  

separate comment below regarding this feature / fix

derammo commented:
regarding this feature and whether you want to take it:

Learning Helios code and implementing this fix took about half an hour.  The next day I made it into a prettier feature with configuration.   So I am not heavily invested and not at all offended if you don't merge it to the product.  I am running off my fork, so I can just keep using this feature merged into whatever you end up releasing.

derammo commented:
PS: also, I will continue to offer up other work even if you don't take this one :)

BlueFinBima commented:
Hi @derammo my only motivation is to ensure that, where possible, we are fixing the the root cause of a problem and that we make Helios better.  Very thankful for all the background you provided.  This still smells like there is something fishy going on here, but since this is a problem others have hit, your code represents a pragmatic solution.  Helios's deserializers are not clever so I want to do some more work to ensure that profiles written with this feature can be loaded by older control centers.  This is being done in branch "issue122".  I'm going to leave this pull request open for now.
I'm very happy to have future contributions from you.  I'm also happy to extend an invitation to the project if you're interested.

derammo commented:
@BlueFinBima if you find the setting in Windows 10 that makes this behavior go away, let me know. 

Either way, I am happy to have this feature eventually.  I think that working correctly in a default configuration is goodness.  If possible, we should let people use their touch screens for other things besides Helios, without having to reconfigure.  

I wish I had found a way to more cleanly ignore all "gestures" that might be configured on a user's machine.  But it appears that these just generate mouse events and we can't just ignore mouse interaction.  Obviously, people use mice and trackballs to click stuff on their Helios displays.  So for now, correlation based on time of the event is the best I came up with.  Maybe there is something in the event structures that can correlate more cleanly?   Also interesting whether this is simply Windows' touch screen driver doing it, or whether there will be differences across different device makers.   So the pragmatic optional thing we have now might be safest.

derammo commented:
follow up: this enhancement is not really needed.   the touch screen continues to make extra events but nothing bad happens since I eliminated the circularities in the profile.   So it does in fact create a bunch of extra events, but they don't propagate to DCS to cause any trouble.  Perhaps there is some temporal filtering on these events somewhere.

derammo commented:
I stand corrected.  I still need this fix.   The problems are back with my touch panels, running from Dev branch.  I may have been running with my private branch when I thought it was fixed.   So I will be submitting this again as a pull request soon, changed in such a way that it does not break profile compatibility.

derammo commented:
new pull request will come when this code is on the right branch against Dev

BlueFinBima commented:
Replaced the per-monitor per-profile settings with a Control Center preference which will be a global for all of Helios's monitors.  Disabling code remains the same.
